daemons/pacemakerd/pacemakerd.c

```diff
@@ -1170,7 +1248,13 @@ main(int argc, char **argv)
     mainloop_add_signal(SIGTERM, pcmk_shutdown);
     mainloop_add_signal(SIGINT, pcmk_shutdown);
 
-    init_children_processes();
+    if (running_with_sbd) {
+        pacemakerd_state = XML_PING_ATTR_PACEMAKERDSTATE_WAITPING;
```

Comment:
I forgot to mention earlier, we need to log a notice that we're waiting on sbd at this point. If something goes wrong and sbd doesn't ping, users could be really confused as to why nothing is happening. Probably also worthwhile to log a message when starting everything after getting the ping.

Reply:
Don't the daemons have enough logging when they come up?
Logging when we start waiting is of course a good idea.

Reply:
Just thinking that people will want to match "waiting for sbd connection" to "received sbd connection"

Reply:
On shutdown we'd like some notice as well when we are waiting for the completed shutdown to be picked up.

Reply:
There's a special case we didn't think of though. With SBD_SYNC_RESOURCE_STARTUP enabled, either explicitly or by default, it's also possible that meanwhile even pacemaker integration is disabled on sbd side. So startup here could be hanging forever. 

And AFAICT, there seems to be difficulty for pacemaker to detect that exactly, given that pacemaker integration could be disabled either with SBD_PACEMAKER=no or "-PP" option...

Do you think this could be somehow made smarter, or should we consider it's a misconfiguration :-) ?

Reply:
Or probably we should at least handle the case of SBD_PACEMAKER=no the smarter way, and leave the case of "-PP" to users and document it better in sbd man page.

Reply:
That sounds reasonable to me

Reply:
Hmm ... different handling of options coming in via environment vs. cmdline might create even more confusion.
Not adding a cmdline option for SBD_SYNC_RESOURCE_STARTUP was kind of intentional.
And of course I would like to not pull too much of the logic/code into pacemaker.
To get the configuration running it would still be possible to switch off SBD_SYNC_RESOURCE_STARTUP
together with disabling pacemker-awareness in  the config-file.
So what we're talking about is to keep an existing config running without having to change the config-file on update - right?
Guess we don't want to share a library between sbd & pacemkaker - wrong direction of dependencies atm anyway.
Other approach might be adding a query to sbd-cmdline ... maybe together with locating the daemon or some pre-flight-stuff.

Reply:
It definitely makes sense to try keeping an existing configuration working after update if possible. 

Generally, with pacemaker integration disabled,  we probably could argue that an explicit setting SBD_SYNC_RESOURCE_STARTUP=no is a misconfiguration, which would require users to fix by themselves. But with SBD_SYNC_RESOURCE_STARTUP omitted in the configuration while it defaults to "yes", cluster refuses to start, which still requires users to fix by themselves by explicitly setting SBD_SYNC_RESOURCE_STARTUP=no. That's probably hard to be acceptable for users...

But of course I can tell how tricky it is to make sbd and pacemaker consider the cases consistently given the current facts...

I don't have a better idea without introducing more complicated mechanism for this for the timing being. If we wouldn't want pacemaker to check and rely on SBD_PACEMAKER, AFAICT the least thing probably we could do now is to tell users what this is about with an error message, ask them to change the configuration and let sbd startup fail already, rather than letting users wait forever for pacemaker to start up. What do you say?

Reply:
I was thinking to directly fail the start of sbd upon paring of the configuration in that case, before pacemaker even gets the chance to start... 

But in case we wouldn't want to change any current behavior, at least dropping an error log could be helpful. Users then would need to shutdown cluster services by themselves and fix the configuration. Currently, that is not even possible either since pacemaker will get stuck again upon shutdown on " Waiting for SBD to pick up shutdown-complete-state".